### PR TITLE
COOPR-685 Fix leaking ZooKeeper watches

### PR DESCRIPTION
**Pull request by jwang47**
Two major issues that this PR addresses:
- Better distributed lock that doesn't leak watches
- Timeout for requests to provisioners

### CODE REVIEW

File: coopr-server/src/main/java/co/cask/coopr/common/zookeeper/lib/ReentrantDistributedLock.java [hunk 1]
```diff
@@ -0,0 +1,402 @@
...
+import java.util.concurrent.locks.Lock;
+import java.util.concurrent.locks.ReentrantLock;
+import javax.annotation.Nullable;
+
+/**
+ * A reentrant distributed lock implementation that uses ZooKeeper using the receipt described in
+ *
+ * http://zookeeper.apache.org/doc/trunk/recipes.html#sc_recipes_Locks
+ */
+public final class ReentrantDistributedLock implements Lock {
```
Comment (albertshau) [inline coopr-server/src/main/java/co/cask/coopr/common/zookeeper/lib/ReentrantDistributedLock.java:53]: is this taken from cdap-common? If so, should open a jira to move that to the common repo.

**Comment by albertshau** [comment]
:+1:

File: coopr-server/src/main/java/co/cask/coopr/common/zookeeper/lib/ReentrantDistributedLock.java (hunk 1 above)
Comment (jwang47) [inline coopr-server/src/main/java/co/cask/coopr/common/zookeeper/lib/ReentrantDistributedLock.java:53]: Created https://issues.cask.co/browse/CDAP-1468

**Comment by jwang47** [comment]
Maven build passed, so merging.

**Pull request merged**